### PR TITLE
Add (optional) code style check before commits (w/pre-commit + Black)

### LINKED ISSUE #85: Installing from source in editable mode adds an empty directory and a shared library (.so) file
#### Describe the bug

Installing from source in editable mode `pip install -e .[dev]` adds a `home/<path_to_project_root_directory>/rsciio/tests/bruker_data/` directory with a file `test_compilers.o` and a `rsciio/bruker/unbcf_fast.cpython-310-x86_64-linux-gnu.so` file to the project root directory (i.e. the `home/<path...` starts in the project root directory). I assume these are created when building Cython extensions, but I don't know anything about that, I'm afraid.

#### To Reproduce

```bash
# No changes
> git status
On branch remove-persitent-search-field
Your branch and 'origin/remove-persitent-search-field' have diverged,
and have 16 and 3 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)

nothing to commit, working tree clean

> pip install -e .[dev]
Obtaining file:///home/hakon/kode/rosettasciio
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
[...]
Successfully built RosettaSciIO
Installing collected packages: RosettaSciIO
  Attempting uninstall: RosettaSciIO
    Found existing installation: RosettaSciIO 0.1.dev0
    Uninstalling RosettaSciIO-0.1.dev0:
      Successfully uninstalled RosettaSciIO-0.1.dev0
Successfully installed RosettaSciIO-0.1.dev0

# Two files added (one inside a directory)
> git status
On branch remove-persitent-search-field
Your branch and 'origin/remove-persitent-search-field' have diverged,
and have 16 and 3 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	home/
	rsciio/bruker/unbcf_fast.cpython-310-x86_64-linux-gnu.so

nothing added to commit but untracked files present (use "git add" to track)
```

#### Expected behavior

Expected a clean repository after installing dev resources from source.

#### Python environement:
 - RosettaSciIO version: 0.1.dev
 - Python version: 3.10.9

### PR DESCRIPTION
**Pull request by hakonanes**
### Description of the change
I've added an option for developers to run a code style check before commiting with git (a git hook), using [pre-commit](https://pre-commit.com/) and Black (see Black's [Version control integration page](https://black.readthedocs.io/en/stable/integrations/source_version_control.html)). These are added to the development dependencies.

The "Lint" GitHub check in this repo uses the (latest) stable version of Black to check changes. pre-commit requires specific versions of software used in their git hooks ([reasoning here](https://pre-commit.com/#using-the-latest-version-for-a-repository)). I set it to the latest, v23.1.0, however the "Lint" check and the pre-commit version might become out of sync at some point. When that happens, a pre-commit config update can be run with `pre-commit autoupdate` (and the updated config file must be pushed to `main`).

The use of this optional code style check should be added to the contributing guide. However, I noticed that in the relevant section, RosettaSciIO's contributing refers to HyperSpy's contributing guide. Should I add a brief explanation below this reference, or should all this be done in HyperSpy as well?

Also updated .gitignore to fix #85.

### Progress of the PR
- [x] add an changelog entry in the `upcoming_changes` folder (see [`upcoming_changes/README.rst`](https://github.com/hyperspy/rosettasciio/blob/main/upcoming_changes/README.rst)),
- [x] Check formatting changelog entry in the `readthedocs` doc build of this PR (link in github checks)
- [ ] Update contributing guide
- [ ] ready for review.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,9 +2,10 @@
...
+home/
 *__pycache__*
 *test_compilers.obj
 *pyd
 .spyproject/*
-rsciio/bruker/unbcf_fast.c
-.idea
\ No newline at end of file
+rsciio/bruker/unbcf_fast.c*
+.idea
```
Comment (ericpre) [inline .gitignore:11]: How do these changes filter the `unbcf_fast*.so` file?

File: .gitignore [hunk 2]
```diff
@@ -2,9 +2,10 @@
...
 /RosettaSciIO.egg-info/*
+home/
 *__pycache__*
 *test_compilers.obj
 *pyd
 .spyproject/*
-rsciio/bruker/unbcf_fast.c
-.idea
\ No newline at end of file
+rsciio/bruker/unbcf_fast.c*
```
Comment (ericpre) [inline .gitignore:10]: Why is this change necessary?

Comment (hakonanes) [reply .gitignore:10]: Without the wildcard, the `rsciio/bruker/unbcf_fast.cpython-310-x86_64-linux-gnu.so` file produced from `pip install -e .[dev]` is not ignored by git.

File: .gitignore (hunk 1 above)
Comment (hakonanes) [reply .gitignore:11]: Right, this is a directory used in PyCharm projects. Should I add it in a separate commit?

File: .gitignore [hunk 3]
```diff
@@ -2,9 +2,10 @@
 /build/*
 /dist/*
 /RosettaSciIO.egg-info/*
+home/
 *__pycache__*
 *test_compilers.obj
```
[the remark this answers is not in the document]
Comment (hakonanes) [reply .gitignore:7]: Good point, much better. Thank you.

File: setup.py [hunk 4]
```diff
@@ -196,7 +196,10 @@ def run(self):
 }
 extras_require["all"] = list(itertools.chain(*list(runtime_extras_require.values())))
 
-extras_require["dev"] = list(itertools.chain(*list(extras_require.values())))
+extras_require["dev"] = [
+    "black",
+    "pre-commit",
```
[the remark this answers is not in the document]
Comment (hakonanes) [reply setup.py]: Understandable, I've removed it.

File: upcoming_changes/86.maintenance.rst [hunk 5]
```diff
@@ -0,0 +1,2 @@
+Add black and pre-commit as development dependencies.
```
[the remark this answers is not in the document]
Comment (hakonanes) [reply upcoming_changes/86.maintenance.rst]: Thanks, I've updated the changelog entry.

File: .gitignore (hunk 2 above)
Comment (ericpre) [reply .gitignore:9]: I see and even if this is working, I would suggest to use the pattern suggested in https://github.com/hyperspy/rosettasciio/issues/85#issuecomment-1419706822, because they map the pattern to the file type to be filtered more explicitly:
There are two types of be ignore:
- the c code `unbcf_fast.c`
- and the corresponding binary `unbcf_fast.so` on unix or `unbcf_fast.pyd` on windows

I don't think we should depend on the way python name the compiled file (`unbcf_fast.cpython-310-x86_64-linux-gnu`) as the `cpython` part could change easily but the file extension (`so` or `pyd`) should be stable.

File: .gitignore (hunk 1 above)
Comment (ericpre) [reply .gitignore]: No this is fine, I was asking about the `unbcf_fast*.so` file and this is discussed below.

File: .gitignore (hunk 2 above)
Comment (hakonanes) [reply .gitignore]: I've used the pattern you refer to. So sorry for missing this in the first place.

**Review by ericpre — APPROVED** [review]
